source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,239 @@
+        if (message_.result_id_map().contains(param->result_id())) {
+          found_entry_for_parameter = true;
+        }
+      });
+  return !found_entry_for_parameter;
+}
+
+void TransformationInlineFunction::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  auto caller_function =
+      ir_context->get_instr_block(function_call_instruction)->GetParent();
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+
+  for (auto& block : *called_function) {
+    // The called function entry block label will not be inlined.
+    if (&block != &*called_function->entry()) {
+      auto cloned_label_instruction = block.GetLabelInst()->Clone(ir_context);
+      assert(message_.result_id_map().contains(
+                 cloned_label_instruction->result_id()) &&
+             "Label id must be mapped to a fresh id.");
+      uint32_t fresh_id =
+          message_.result_id_map().at(cloned_label_instruction->result_id());
+      cloned_label_instruction->SetResultId(fresh_id);
+      function_call_instruction->InsertBefore(
+          std::unique_ptr<opt::Instruction>(cloned_label_instruction));
+      fuzzerutil::UpdateModuleIdBound(ir_context, fresh_id);
+    }
+
+    for (auto& instruction : block) {
+      // Replaces the operand ids with their mapped result ids.
+      auto cloned_instruction = instruction.Clone(ir_context);
+      cloned_instruction->ForEachInId(
+          [this, called_function, function_call_instruction](uint32_t* id) {
+            // If the id is mapped, then set it to its mapped value.
+            if (message_.result_id_map().count(*id)) {
+              *id = message_.result_id_map().at(*id);
+              return;
+            }
+
+            uint32_t parameter_index = 0;
+            called_function->ForEachParam(
+                [id, function_call_instruction,
+                 &parameter_index](opt::Instruction* parameter_instruction) {
+                  // If the id is a function parameter, then set it to the
+                  // parameter value passed in the function call instruction.
+                  if (*id == parameter_instruction->result_id()) {
+                    *id = function_call_instruction->GetSingleWordInOperand(
+                        parameter_index + 1);
+                  }
+                  parameter_index++;
+                });
+          });
+
+      // If |cloned_instruction| has a result id, then set it to its mapped
+      // value.
+      if (cloned_instruction->HasResultId() &&
+          message_.result_id_map().count(cloned_instruction->result_id())) {
+        assert(message_.result_id_map().contains(
+                   cloned_instruction->result_id()) &&
+               "Result id must be mapped to a fresh id.");
```

Comment:
The assertion will never fail.

Reply:
It will fail if some instruction, inside `called_function`, is not mapped.

Reply:
I've just noticed that condition in the `if` statement is the same (almost) as the one in the assertion (I believe `count` function returns `0` if `contains` returns `false`).